channel/fd_channel.cpp

```diff
@@ -0,0 +1,161 @@
+#include "fuzzuf/channel/fd_channel.hpp"
+#include "fuzzuf/logger/logger.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+#include <signal.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+FdChannel::FdChannel() {
+    // Nothing to do
+}
+
+FdChannel::~FdChannel() {
+    TerminateForkServer();
+}
+
+int FdChannel::Send(void *buf, size_t size) {
+    int nbytes = write(forksrv_write_fd, buf, size);
+    if (nbytes < 0) {
+        ERROR("[FdChannel] Failed to send");
+    }
+    return nbytes;
+}
+
+// Recieved data to be stored to user allocated pointer `buf`
+int FdChannel::Recv(void *buf, size_t size) {
+    int nbytes = read(forksrv_read_fd, buf, size);
+    if (nbytes < 0) {
+        ERROR("[!] [FdChannel] Failed to recv");
+    }
+    return nbytes;
+}
```

Comment:
1. The type of `nbytes` should be `ssize_t`. You can ignore this if you're sure the size is always less than 0x80000000.
2. As pointed out by @potetisensei for other codes, `read` and `write` might fail to read or write data of the specified size. There are multiple possibilities those system calls result in failure.
At first, `read`/`write` may return a positive value even if it couldn't actually read/write the data. This happens when the size is too large or the sender/reader splits the data due to a buffering and so on. You must write a loop until it successfully reads/writes the whole data.
Secondly, these system calls may fail even if the data is available. In this case it returns a negative value (-1) and sets the reason to `errno`. This happens because of file blocking problems and signals made during reading/writing data. You should check `errno` and retry in case the data is available.

Although I'm not sure if checking `errno` is necessary in this case, I believe using a loop is desired to prevent some bugs in the future. Here is my suggestion: (Note you need to change `fd_channel.hpp` too)
```suggestion
size_t FdChannel::Send(void *buf, size_t size) {
    size_t nbytes = 0;

    do {
        ssize_t n = write(forksrv_write_fd, buf, size);
        if (n < 0) {
            if ((errno != EINTR) && (errno != EAGAIN) && (errno != EWOULDBLOCK)) {
                ERROR("[FdChannel] Failed to send");
            } else {
                n = 0;
            }
        }

        nbytes += n;
        size -= n;
    } while (size);

    return nbytes;
}

// Recieved data to be stored to user allocated pointer `buf`
size_t FdChannel::Recv(void *buf, size_t size) {
    size_t nbytes = 0;

    do {
        ssize_t n = read(forksrv_read_fd, buf, size);
        if (n < 0) {
            if ((errno != EINTR) && (errno != EAGAIN) && (errno != EWOULDBLOCK)) {
                ERROR("[FdChannel] Failed to send");
            } else {
                n = 0;
            }
        }

        nbytes += n;
        size -= n;
    } while (size);

    return nbytes;
}
```

Reply:
Also, I prefer the design to return -1 on failure instead of calling `ERROR`.
If you do so, you can use `FdChannel::Recv` in `FdChannel::WaitForkServerStart`, for example.

Reply:
You can resolve all of the problems above by calling `read_n` of `utils/common.cpp`.

Reply:
Thank you for suggestion. 
I fixed Recv() and Send() to use ReadFile() and WriteFile().
These function tries to read/write n bytes exactly.
Resolved.